Initial implementation of helper library for Rust UDFs

wmitros opened the pull request.

piodul commented on bindgen/src/lib.rs:
```diff
@@ -0,0 +1,238 @@
+}
+
+#[proc_macro_attribute]
+pub fn scylla_bindgen(attrs: TokenStream, item: TokenStream) -> TokenStream {
```
The function is very large. I'd recommend splitting it up.

wmitros replied:
It should be more readable now

piodul commented on bindgen/src/lib.rs:
```diff
@@ -0,0 +1,238 @@
+    ];
+    let native_types = vec!["u32", "u64", "i32", "i64", "f32", "f64"];
+    let parsed_item = parse::<ItemFn>(item);
+    if let Err(_e) = parsed_item {
+        return TokenStream::from(
+            syn::Error::new(
+                Span::call_site(),
+                "scylla_bindgen can only be used on functions.",
+            )
+            .to_compile_error(),
+        );
+    }
+    let input = parsed_item.unwrap();
```
Nit: I'd recommend against using unwraps when they can be safely avoided, e.g. via pattern matching. Here, you can use:

```rust
let input = match parsed_item {
    Ok(item) => item,
    Err(_) => return TokenStream::from(
        syn::Error::new(
            Span::call_site(),
            "scylla_bindgen can only be used on functions.",
        )
        .to_compile(_error()
    ),
};
```

wmitros replied:
Ok

piodul commented on serialization/src/lib.rs:
```diff
@@ -0,0 +1,185 @@
+}
+
+pub trait IntoCQL {
+    fn as_cql(&self) -> u64;
```
For increased type safety, I think you could introduce a type-safe wrapper over `u64` which represents a handle to a CQL value. It should be marked with `#[repr(transparent)]` so that it has exactly the same layout as a bare `u64` and is treated by the ABI in the same way.

wmitros replied:
I added a `SizePointer`

piodul reviewed (changes requested):
Some general tips on how the code structure could be improved:

The `scylla_bindgen` function is very complex, has lots of variables and operates directly on AST which is a bit hard to parse. It definitely needs to be broken down into smaller components. It's not really clear at the first glance what is the general idea behind the transformation it performs.

In order to improve readability and maintainability, I suggest introducing an intermediate representation - an abstract model that would describe the user-supplied function in such a way that generating code later is easy.

For example (I'm making this up, might not compile):

```rust
struct UserDefinedFunctionDesc {
	arguments: Vec<ArgumentDesc>,
	return_type: ReturnTypeDesc,

	original_function_name: syn::Ident,
	original_body: syn::Block,
}

struct ArgumentDesc {
	name: syn::Ident,
	typ: syn::Type,
}

struct ReturnType {
	typ: syn::Type,
}
```

First, you could build the model:

```rust
// Purposefully avoiding error handling here for simplicity
impl UserDefinedFunctionDesc {
	fn new(attrs: TokenStream, input: TokenStream) -> Self {
		let func = parse::<ItemFn>(item).unwrap();
		let arguments = func.sig.inputs.iter().map(ArgumentDesc::new).collect();
		let return_type = ReturnType::new(func.sig.output);

		// ... more stuff ...

		Self {
			arguments,
			return_type,

			original_function_name,
			original_body,
		}
	}
}
```

...and then synthesize the code based on it:

```rust
impl UserDefinedFunctionDesc {
	fn generate(&self) -> TokenStream {
		let exported_function = self.generate_exported_function();
		let inner_function = self.generate_inner_function();

		quote! {
			#exported_function
			#inner_function
		}
	}

	fn generate_exported_function(&self) -> syn::ItemFn { /* ... */ }
	fn generate_inner_function(&self) -> syn::ItemFn { /* ... */ }
}

impl ArgumentDesc {
	fn exported_arg_type(&self) -> syn::Argument { /* ... */ }
	fn inner_arg_type(&self) -> syn::Argument { /* ... */ }
}

// etc. etc.
```

wmitros commented on scylla-cql/src/frame/response/cql_to_rust.rs:
```diff
@@ -96,6 +96,7 @@ impl_from_cql_val!(IpAddr, as_inet); // IpAddr::from_cql<CqlValue>
 impl_from_cql_val!(Uuid, as_uuid); // Uuid::from_cql<CqlValue>
 impl_from_cql_val!(BigDecimal, into_decimal); // BigDecimal::from_cql<CqlValue>
 impl_from_cql_val!(Duration, as_duration); // Duration::from_cql<CqlValue>
+impl_from_cql_val!(CqlDuration, as_cql_duration); // CqlDuration::from_cql<CqlValue>
```
I submitted https://github.com/scylladb/scylla-rust-driver/pull/623

wmitros commented on bindgen/src/lib.rs:
```diff
@@ -0,0 +1,238 @@
+        #atrs
+        #[derive(UDTToColumnType)]
+        #[derive(FromUserType)]
+        #[derive(IntoUserType)]
```
Ok

wmitros commented on bindgen/src/lib.rs:
```diff
@@ -0,0 +1,238 @@
+                Box::new(syn::Type::from(
+                    parse_str::<syn::TypePath>("u64").expect("code generation error"),
+                )),
+            );
```
I'm using `quote!` almost everywhere now

wmitros commented on bindgen/src/lib.rs:
```diff
@@ -0,0 +1,238 @@
+}
+
+#[proc_macro_attribute]
+pub fn scylla_bindgen(attrs: TokenStream, item: TokenStream) -> TokenStream {
```
I added an example.

piodul replied:
The example was put into a docstring of a public item. I don't think it's a good idea because some users might think that we are guaranteeing that the generated code will always look like this, while it's an implementation detail that might change.

My intention was just to provide some guidance for the future developers of the library. I would recommend moving that docstring to `export_udf::export_udf` in order to hide it from the users' eyes. Alternatively, we could drop it as I think the code is now much easier to understand than before, and people have a tendency to forget to update comments so it may just increase confusion instead of reducing it in the future.

wmitros replied:
I moved the example to `export_udf::export_udf` and added some more comments in its place

wmitros commented on bindgen/src/lib.rs:
```diff
@@ -0,0 +1,238 @@
+                id.mutability = Option::None;
+            }
+        }
+    }
```
I went ahead with this approach in the rework

wmitros commented on bindgen/src/lib.rs:
```diff
@@ -0,0 +1,238 @@
+    for inp in &mut sig_exported.inputs {
+        if let &mut FnArg::Typed(ref mut pat) = inp {
+            let typestr = pat.ty.to_owned().into_token_stream().to_string();
+            if !scalar_types.iter().any(|&s| s == typestr) {
```
I abandoned the idea of comparing strings for type matching in the rework - now there's a `WasmConvertible` trait for it

wmitros commented on examples/topn_row.rs:
```diff
@@ -0,0 +1,16 @@
+use std::collections::BTreeSet;
+#[scylla_bindgen]
+fn topn_row(
+    acc: (i32, BTreeSet<(i32, String, i32)>, i32),
```
I did use this, the first methods does work in the rework

wmitros commented on examples/topn_row.rs:
```diff
@@ -0,0 +1,16 @@
+    acc: (i32, BTreeSet<(i32, String, i32)>, i32),
+    v: String,
+) -> (i32, BTreeSet<(i32, String, i32)>, i32) {
+    let mut acc = acc;
```
Ok

wmitros commented:
I addressed most of the comments (though not all yet) and added a few more features:
 - restructuring the entire code, it should look more like a proper crate now
 - allowing any pattern as a function parameter
 - adding a trait for types that can be an arguments of the Rust UDF and using it to get the correlating types for the WASM UDF instead of matching strings - this allows for example using aliases
 - reexporting everything that we're using in macros in `_macro_internal` and using the full path in the macros (didn't add renaming of the `scylla_bindgen` crate yet)
 - adding a macro for `newtype` idiom, so that the "newtypes" can be used as parameters in the UDFs if the underlying type can be used there

piodul commented on bindgen/src/lib.rs:
````diff
@@ -0,0 +1,84 @@
+/// ```
+/// SELECT foo(x) FROM table;
+/// ```
+pub use bindgen_macros::scylla_newtype;
````
I'm not convinced about the usefulness of this macro. I would imagine that users would create their newtype wrappers in order to implement serialization/deserialization _differently_ to how the trait impls for the inner type do it, so what this macro does is contrary to the idea of newtype wrappers.

wmitros replied:
I added this mostly to enable users to overwrite traits of types that represent cql values without having to rewrite the serialization. For simple types, this could be relatively simply avoided by just creating the newtype from the cql argument, but it becomes problematic with complex types - you need to reconstruct the entire structure then.
While recreating the complex value does not increase the computational complexity of the function, it may almost double it, if most of the work of the function is deserialization, and this is something we probably would still like to avoid.
An example use case for this is in `examples/topn_row.rs`. We store the top k values using an ordered set there, and thanks to this macro we avoid reimplementing the serialization of strings or the rebuilding of the accumulator set.

piodul replied:
Ok, I see now. It is useful when the newtype is a part of a more complex type, e.g. an element of a set, like in your example.

piodul commented on bindgen-macros/src/udt.rs:
```diff
@@ -0,0 +1,75 @@
+
+pub(crate) fn user_defined_type(_attrs: TokenStream, item: TokenStream) -> TokenStream {
+    let mut st = syn::parse_macro_input!(item as syn::ItemStruct);
+    // We have to make the struct public, otherwise we can't use "<$struct_name as WasmConvertible>::WasmType"
+    // in the exported function signature even if WasmType is public.
```
It might be a bit surprising for the user to find out that their struct is made public, even though they didn't mark it that way.

Is the error message clear in case the struct is not `pub`, i.e. tells that the problem is caused by the struct having wrong visibility and explains why it is needed? If so then I think it's better to inform the user and let them make the fix themselves.

wmitros replied:
In the rebase, the exported UDF function doesn't even have to be `pub` so we also don't have to concern ourselves with this.

piodul replied:
I can see that there is still a similar issue in the `export_newtype` macro.

wmitros replied:
Actually, I just forgot to remove the comment there.

piodul commented on bindgen-serialization/src/wasmconvertible.rs:
```diff
@@ -0,0 +1,159 @@
+impl_wasm_convertible_scalar!(i16, i32);
+impl_wasm_convertible_scalar!(i32, i32);
+impl_wasm_convertible_scalar!(i64, i64);
+impl_wasm_convertible_scalar!(u8, u32);
+impl_wasm_convertible_scalar!(u16, u32);
+impl_wasm_convertible_scalar!(char, u32);
```
Should UDFs really support signed types? There is no notion of signed integers in CQL.

wmitros replied:
We probably shouldn't. I removed them (I assume you meant **un**signed)

piodul replied:
Yeah, I meant unsigned.

piodul commented on bindgen-serialization/src/wasmconvertible.rs:
```diff
@@ -0,0 +1,159 @@
+
+macro_rules! impl_wasm_convertible_sizeptr_tuple {
+    ( $( $types:ident )* ) => {
+        impl<$($types),*> WasmConvertible for ($($types,)*) where
+        ($($types,)*) : FromCQL + AsCQL {
```
Nit: please format the `while` similarly to what you did in previous `macro_rules!`:
```rust
impl<$($types),*> WasmConvertible for ($($types,)*)
where
    ($($types,)*) : FromCQL + AsCQL,
{
```

wmitros replied:
Ok

piodul commented on bindgen-macros/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+
+[dependencies]
+quote = "1.0.15"
+syn = { version = "1.0.86", features = ["full", "visit-mut"] }
```
Those features shouldn't be needed now, right? At least `visit-mut` should be unnecessary.

wmitros replied:
It works without them now

piodul commented on examples/topn_final.rs:
```diff
@@ -0,0 +1,10 @@
+use scylla_bindgen::scylla_bindgen;
+use std::collections::BTreeSet;
```
Nit: leave one empty line after imports

wmitros replied:
Ok

wmitros commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -558,7 +558,7 @@ fn deser_prepared_metadata(buf: &mut &[u8]) -> StdResult<PreparedMetadata, Parse
 }
 
-fn deser_cql_value(typ: &ColumnType, buf: &mut &[u8]) -> StdResult<CqlValue, ParseError> {
+pub fn deser_cql_value(typ: &ColumnType, buf: &mut &[u8]) -> StdResult<CqlValue, ParseError> {
```
I submitted https://github.com/scylladb/scylla-rust-driver/pull/630

wmitros commented on serialization/src/lib.rs:
```diff
@@ -0,0 +1,185 @@
+        let size = (sizeptr >> 32) as usize;
+        let ptr = (sizeptr & 0xffffffff) as *mut u8;
+        let mut slice = unsafe { std::slice::from_raw_parts(ptr, size) };
+        T::from_cql(deser_cql_value(&T::to_column_type(), &mut slice).ok()).unwrap()
```
I applied your recommendations after confirming that we're treating `Option<T>` the same in the driver and here

wmitros commented on serialization/src/lib.rs:
```diff
@@ -0,0 +1,185 @@
+
+impl<T> FromCQL for Option<T>
+where
+    T: FromCqlVal<Option<CqlValue>> + ToColumnType,
```
I removed this case in the rebase, but I added `impl<T: ToColumnType> ToColumnType for Option<T>` to handle nulls appearing in tuples

wmitros commented on bindgen/src/lib.rs:
````diff
@@ -0,0 +1,84 @@
+/// ```
+pub use bindgen_macros::user_defined_type;
+
+/// The macro takes a "newtype" struct (tuple struct with only one field) and generates all implementations for traits used in the the
````
Ok

wmitros commented on bindgen-macros/src/udf.rs:
```diff
@@ -0,0 +1,115 @@
+    // We only need to declare the ABI once.
+    static mut DECLARED_ABI: std::sync::atomic::AtomicBool =
+        std::sync::atomic::AtomicBool::new(false);
+    if !unsafe { DECLARED_ABI.swap(true, std::sync::atomic::Ordering::Relaxed) } {
```
The compiler complained when it wasn't unsafe even with the atomic variable, but it got silenced when I removed the `mut` from the static - interestingly, we can still perform `swap` on a non-mut static

piodul replied:
AFAIK using global mutable variables requires unsafe. However, atomics are ones of the "magic" types in Rust that you can modify through immutable access.

wmitros commented on bindgen-macros/src/udf.rs:
```diff
@@ -0,0 +1,115 @@
+
+fn get_exported_fun(
+    item: &ItemFn,
+    parameters: &Vec<TokenStream2>,
```
Ok

wmitros commented on bindgen-serialization/src/ascql.rs:
```diff
@@ -0,0 +1,35 @@
+            if siz == u32::MAX {
+                return size_ptr(siz, 0);
+            }
+            let dest = _scylla_malloc(siz as usize);
```
It does, I added handling for that

wmitros commented on bindgen-serialization/src/ascql.rs:
```diff
@@ -0,0 +1,35 @@
+use crate::{SizePointer, _scylla_malloc, size_ptr};
+use scylla_cql::frame::value::Value;
+
+pub trait AsCQL {
```
I changed `as` to `to`

wmitros commented on bindgen-serialization/src/wasmconvertible.rs:
```diff
@@ -0,0 +1,159 @@
+    };
+}
+
+// impl_wasm_convertible!(Vec<u8>, SizePointer);
```
I was tracking that all cql types were implemented, and this (impl for blobs) was already implemented by the generic impl for `Vec` (lists). I added a comment about it so that it isn't forgotten

wmitros commented on bindgen-macros/src/udf.rs:
```diff
@@ -0,0 +1,115 @@
+fn get_abi() -> TokenStream2 {
+    // We only need to declare the ABI once.
+    static mut DECLARED_ABI: std::sync::atomic::AtomicBool =
+        std::sync::atomic::AtomicBool::new(false);
```
I tried a different approach in the rebase - the `_scylla_abi` static is now exported directly in the crate, and not in the macro.

I exported it with `#[no_mangle]` and `#[used]` attributes, but in theory it may not be enough to make it end up in the final wasm binary. This would be achieved using the `#[used(linker)]` unstable feature, but if I'm not mistaken, `#[used]` guarantees that it reaches the linking stage, and we can make sure it doesn't get stripped during linking by referencing it in a public function.
We could reference it in the UDF, but that would make it (slightly) slower. We could also create a new function that does reference it - that makes the binary (slightly) bigger - I went with that for now, because for a very simple function that only performs some calculations the cost of accessing the global seems proportionally higher than the bigger executable, which will always be quite sizeable because it contains the implementation of `malloc`. 

Note that, for now, the function is actually not needed - for non-ELF targets the `#[used]` attribute does all that we want (in fact, even `#[no_mangle]` is enough), but in theory it's not guaranteed.
I'm also not against adding a reference to the original generated function implementation, if that would be good enough

wmitros replied:
It turns out that neither `#[used]` nor `pub` have any useful effect in terms of exporting the symbol in the final binary. At least in the current Rust version, we have to tag a item with `#[no_mangle]` or `#[export_name = ...]` and it will be exported regardless of whether it was placed in the root crate or whether it was even `pub`.
However, this behavior is not guaranteed to continue, but I found no immediate plans of changing it.
Because of that, I suggest that we depend on it for now. It has 2 main benefits: we don't have to make implementation hacks to make it work, and we don't have to require users to be extra careful in terms of placement of the items exported to Scylla and their visibility setting.

piodul replied:
> At least in the current Rust version, we have to tag a item with `#[no_mangle]` or `#[export_name = ...]` and it will be exported regardless of whether it was placed in the root crate or whether it was even `pub`. However, this behavior is not guaranteed to continue, but I found no immediate plans of changing it.

Could you point me to the relevant documentation?

wmitros replied:
The behavior here isn't well documented. We can see that the items survive until the linking phase [here](https://doc.rust-lang.org/reference/abi.html), but the fact that they are also present in the final binary is not guaranteed in general. 
This behavior is actually causing some [issues](https://github.com/rust-lang/rust/issues/73958) because there is no way to prevent exporting an item tagged with `#[no_mangle]` in the final library. The discussion however, is more about how to make it possible to hide such items, and not about changing this default behavior

piodul replied:
We talked about it recently on the meeting and it looks like we should prioritize ease of use. I'm OK with the current approach, especially if it doesn't seem that it will be changed soon.

wmitros commented:
There are still some comments I haven't addressed in the last rebase, but at least all the comments from the previous review are done now.

wmitros commented:
The last rebase fixed a few comments and again changed the structure of the project - this time mostly due to a lot of renames, but also the `bindgen-serialization` package was merged with the original `bindgen` which was practically empty before. Additionally, `Cargo.toml`s have been slightly upgraded. 

The things that weren't fixed in the last rebase are: CI and tests, so I think we can start reviewing this part already.

I'm not actually exposing anything other than the macros and some cql re-exports from `scylla_cql` for now, because I'm not sure the crate is ready for user-defined (de)serialization of types yet, and the traits and types from this crate are only useful for that. I think we will want to make this possible in the future, but I don't think it is needed in the first version.

The other thing that I have to add is a README, but it can also wait until after the code is fixed, and the ability to rename this crate (will be practically the same as what we've added to `scylla_cql`)

wmitros commented on bindgen/src/lib.rs:
```diff
@@ -0,0 +1,84 @@
+    pub use scylla_cql::frame::response::result::ColumnType;
+}
+
+/// The macro transforms a function:
```
I added a line to this and the newtype macro

wmitros commented on serialization/src/lib.rs:
```diff
@@ -0,0 +1,185 @@
+}
+
+pub trait FromCQL {
+    fn from_sizeptr(sizeptr: u64) -> Self;
```
Actually, after I changed the argument from `u64` to a `WasmPtr` can we make this method safe and make the constructor of `WasmPtr` unsafe instead?

piodul replied:
> I made it unsafe in the rebase. I also added a comment about the safety, but I only described that the pointer and size must be valid to create a slice - if the slice does not contain a properly serialized value, the `deser_cql_value` returns an `Err` and this function panics - if I'm not mistaken this behavior is considered safe

Yes, panicking is safe.

> Actually, after I changed the argument from `u64` to a `WasmPtr` can we make this method safe and make the constructor of `WasmPtr` unsafe instead?

Yes, it should be sound to do that.

wmitros replied:
I managed to remove the `unsafe` from `from_sizeptr` by implementing methods for `WasmPtr` better.

piodul commented on scylla-bindgen/src/abi_exports.rs:
```diff
@@ -0,0 +1,25 @@
+/// - the caller must not dereference the returned pointer
+#[no_mangle]
+#[doc(hidden)]
+pub(crate) unsafe extern "C" fn _scylla_malloc(size: u32) -> u32 {
```
Is there a reason why `_scylla_malloc` and `_scylla_free` explicitly use `u32`a instead of pointers/usizes?

wmitros replied:
I'm using `u32` here mainly because it's native for WASM. When Scylla uses these imports, it gets exactly what was returned. It might work with pointers/usizes, but I haven't tested it and I did have some problems with it when compiling C++ to Rust

piodul replied:
Ok, we can keep it that way.

piodul commented on examples/topn_final.rs:
```diff
@@ -0,0 +1,6 @@
+use scylla_bindgen::export_udf;
+
+#[export_udf]
+fn topn_final((_, acc): (i32, std::vec::Vec<String>)) -> std::vec::Vec<String> {
```
Nit: `Vec` is automatically imported, it shouldn't be required to use the full path

wmitros replied:
Ok

piodul commented on examples/return_input.rs:
```diff
@@ -0,0 +1,6 @@
+extern crate scylla_bindgen;
+
+#[no_mangle]
+pub extern "C" fn return_input(wasmptr: u64) -> u64 {
+    wasmptr
+}
```
What is the purpose of this example? It doesn't use the library at all.

It looks like it is supposed to be an identity function which works with all types - but will it? What about types smaller than `u32` (`smallint`, `tinyint`, `int`)?

wmitros replied:
At first I wanted to provide an example that doesn't use the library and results in a much smaller binary, but the example doesn't achieve it anymore and it probably isn't the best place for it, so I'll remove this.

piodul commented on scylla-bindgen-macros/src/export_newtype.rs:
```diff
@@ -0,0 +1,130 @@
+    quote! {
+        impl #impl_generics ::scylla_bindgen::_macro_internal::ToColumnType for #struct_name #ty_generics #where_clause {
+            fn to_column_type() -> ::scylla_bindgen::_macro_internal::ColumnType {
+                #struct_type::to_column_type()
```
Should this call be fully qualified? I.e. `<#struct_type as ::scylla_bindgen::_macro_internal::ToColumnType>::to_column_type()`

wmitros replied:
It probably should, added in rebase

piodul commented:
> > All public items should be documented: traits (`FromCQL`, `AsCQL`, `ToColumnType`, `WasmConvertible`, ...), types (`SizePointer`), maybe more.
> 
> I'm not actually exposing anything other than the macros and some cql re-exports from `scylla_cql` for now, because I'm not sure the crate is ready for user-defined (de)serialization of types yet, and the traits and types from this crate are only useful for that. I think we will want to make this possible in the future, but I don't think it is needed in the first version.

Ok. It looks like those items are not exported because their modules are not exported, so it's fine.

> I'm also not sure if I have to document the `scylla_cql` re-exports

No need, the docstrings in the generated documentation should be inherited from `scylla-cql`.

wmitros commented on bindgen-serialization/src/wasmconvertible.rs:
```diff
@@ -0,0 +1,159 @@
+impl_wasm_convertible_sizeptr_tuple! { A B C D E F G H I }
+impl_wasm_convertible_sizeptr_tuple! { A B C D E F G H I J }
+impl_wasm_convertible_sizeptr_tuple! { A B C D E F G H I J K }
+impl_wasm_convertible_sizeptr_tuple! { A B C D E F G H I J K L }
```
I added some tests. For now, I was only able to test serialization and deserialization by checking if the value that was serialized and then deserialized was still the same value. To check whether the serialized form is correct on its own, I think we could use some integration with Scylla, because the goal is to have the same format as in Scylla.

piodul replied:
I agree, it would be best to check with a real Scylla instance.

I described some of my thoughs thoughts about how we could approach this here: https://github.com/wmitros/scylla-rust-udf/issues/4 . It shouldn't be super hard to set up, but it also requires figuring out some things so it's hard to tell how much time it would actually take (more than one day, for sure).

If we don't do them before releasing to crates, then we should first at least perform some manual tests with existing examples.

wmitros commented on scylla-bindgen/src/wasmptr.rs:
```diff
@@ -0,0 +1,16 @@
+    pub fn raw(&self) -> *mut u8 {
+        (self.0 & 0xffffffff) as *mut u8
+    }
+}
```
I added all of the suggested methods. Additionally, by returning an `Option` whenever necessary, I managed to make all functions safe

wmitros commented:
Looks like I'll have to split your attention again @piodul, as we decided we'll try to finish this before UDF permissions after all.
I addressed all comments and added some tests and a small readme.

wmitros commented:
One additional consideration before publishing the crate is whether we want to change the repo name.
I ended up with the crate name `scylla_bindgen`, but the repo's called `scylla-rust-udf`. I'm fine with keeping the name as is but maybe it's worth to consider something like `scylla-rust-bindgen` (it might be confusing because we also have rust bindings in the core scylla repo) or `scylla-rust-udf-bindgen` (which is very long)

piodul commented:
> One additional consideration before publishing the crate is whether we want to change the repo name. I ended up with the crate name `scylla_bindgen`, but the repo's called `scylla-rust-udf`. I'm fine with keeping the name as is but maybe it's worth to consider something like `scylla-rust-bindgen` (it might be confusing because we also have rust bindings in the core scylla repo) or `scylla-rust-udf-bindgen` (which is very long)

The crates that you are developing are supposed to help with developing UDFs, so I think they should have `udf` in their name. Having `rust` in the crate name is redundant (but it's OK to have it in the repo name).

I think that `scylla-udf` should be OK - or `scylla-udf-bindgen` if you want to be super explicit about the bindgen-like functionality, but IMO that's unnecessarily verbose.

piodul reviewed (commented):
The `wasm_conversion.rs` file is empty? I see that in one of the force-pushes you moved its contents out to another file, but perhaps forgot to delete the old file?

wmitros commented on README.md:
````diff
@@ -1 +1,100 @@
+2. Edit the Cargo.toml to set the crate-type to cdylib
+```
++ [lib]
++ crate-type = ["cdylib"]
````
They were, but on second thought it makes more sense to present the code in a way that's easy to copy. Changed in rebase

wmitros commented on README.md:
````diff
@@ -1 +1,100 @@
+```
+
+
+## CQL Type Mapping
````
I added a line about it

wmitros commented on scylla-bindgen-macros/src/path.rs:
```diff
@@ -0,0 +1,61 @@
+use proc_macro2::TokenStream as TokenStream2;
```
Correct, I removed the `get_path.rs` one

wmitros commented on scylla-bindgen/src/wasmptr.rs:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    pub fn is_null(&self) -> bool {
+        self.size() == None
```
Ok

wmitros commented on scylla-bindgen/src/wasmptr.rs:
```diff
@@ -0,0 +1,77 @@
+        }
+    }
+
+    pub fn null() -> WasmPtr {
```
I made the all `const`

wmitros commented on scylla-bindgen/src/wasm_convertible.rs:
```diff
@@ -0,0 +1,337 @@
+    }
+    #[test]
+    fn counter_convert() {
+        // actually not implemented in scylla
```
I mean the Scylla core. Though after a closer look it looks easy to fix.
The issue is that the have no `serialize` method for a counter in Scylla (https://github.com/scylladb/scylladb/blob/ccc8e47db1807f3cd5088b3cebf0627c1c28dcff/types.cc#L1945), and the UDFs are supposed to return serialized values. As a result, we can't return counters in WASM UDFs. However, when returning, we can treat the counter as an integer, which will fix the issue and which is the approach we're taking for Lua UDFs.

piodul replied:
Ok, thanks for the explanation.

wmitros commented on README.md:
```diff
@@ -1 +1,100 @@
+  For non rustup setups, you can try following the steps at https://rustwasm.github.io/docs/wasm-pack/prerequisites/non-rustup-setups.html
+  Also available as an rpm: `rust-std-static-wasm32-wasi`
+* `wasm2wat` parser
+  Available in many distributions in the `wabt` package
```
I reformatted it

wmitros commented:
> The `wasm_conversion.rs` file is empty? I see that in one of the force-pushes you moved its contents out to another file, but perhaps forgot to delete the old file?

Correct, it should be deleted in the rebase

piodul commented on README.md:
```diff
@@ -1 +1,105 @@
+| BIGINT    | i64                           |
+| BLOB      | Vec\<u8\>                     |
+| BOOLEAN   | bool                          |
+| COUNTER   | scylla_udf::Counter       |
```
Nit: some rows in that table are mis-formatted, please fix.

wmitros replied:
Ok

piodul commented:
I can't get the crate to compile correctly:

```
[piodul@scyllapc scylla-rust-udf]$ cargo test
error: failed to load manifest for workspace member `/home/piodul/code/scylla-rust-udf/examples`

Caused by:
  failed to parse manifest at `/home/piodul/code/scylla-rust-udf/examples/Cargo.toml`

Caused by:
  error reading `dependencies.scylla-bindgen` from workspace root manifest's `workspace.dependencies.scylla-bindgen`

Caused by:
  `dependency.scylla-bindgen` was not found in `workspace.dependencies`
```

I think you missed some places when renaming the crates. Besides, the directory names for the crates are still `scylla-bindgen` and `scylla-bindgen-macros` - please rename them too, as it will be quite confusing if the crate names don't match their directories.

I recommend rebasing the PR on top of main/master, so that the CI has a chance to run.

wmitros commented:
All the cases of scylla-bindgen should be now correctly replaced with scylla-udf. I also made some fixes for rustfmt, clippy and rustdoc, so that CI passes

piodul commented:
I think it looks OK. I have some final requests before we can merge:

- We should document how to run tests. Currently, they don't work when running them natively, and it's better to run them on the target architecture (WASM) anyway. This information will be useful for future contributors.
- Please run some manual tests with the examples in the repo, i.e. compile and optimize them according to the instructions in the README, upload them to a Scylla node, run some queries and verify that they work as expected. Please verify if Scylla correctly serializes/deserializes the types - I think that the types that are passed directly without allocating require the most scrutiny, but please also test that e.g. collections work (if two types have similar implementation then testing only one should be OK). Perhaps we will have some integration tests in the future in this repo, but for a manual check should suffice if we don't plan any significant changes in a followup.
- Please verify that all necessary changes were upstreamed to scylla-cql and scylla-macros (try to compile scylla-udf with that branch).

wmitros commented:
I tested examples with using Scylla's `cql-pytest` suite, the exact test case can be found at https://github.com/wmitros/scylla/commit/91448f1a8283af5584b335dfd0c3859fb03ed7ac

The tests uncovered that one example can't be used with scylla -  the [combine.rs](examples/combine.rs) example took both a counter and a cql duration as arguments, but they cannot be columns of the same table (all non-counter columns in a counter table have to be parts of the partition key, and duration cannot be in a partition key).

The [topn.rs](examples/topn.rs) has also been updated, because the current implementation produced a bit unexpected results - it did return the n longest strings without repetitions, but different strings with the same length were also counted as repetitions. Also, the previous implementation of the row function didn't handle nulls, so if a null row was encountered in the UDA execution, the UDA returned null as well.

I also noticed that the `AS $$ .. $$` workaround described in README doesn't really work, because the WAT code can also have `$$` substrings, so I recommended just using `'`' and doubling the `'` that appear in the WAT code.

wmitros commented:
@piodul is it ok if we merge this before transferring the repo to Scylla?

piodul commented:
@wmitros please squash all the commits (apart from `replace the imported scylla-cql with a git dependency`) and I think we are good to merge.

wmitros commented:
> @wmitros please squash all the commits (apart from `replace the imported scylla-cql with a git dependency`) and I think we are good to merge.

Ok, everything's squashed (I merged `depend on a released scylla-cql` with `replace the imported scylla-cql with a git dependency` because it seemed more correlated)
